src/main/java/com/google/sps/servlets/ChatServlet.java

```diff
@@ -11,66 +11,140 @@
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
-    // Calls query on all entities of type Message
-    Query messageQuery = new Query(MESSAGE_KIND).addSort(TIMESTAMP_PROPERTY, SortDirection.ASCENDING);
-    PreparedQuery preparedMessageQuery = datastore.prepare(messageQuery);
+        // Calls query on all entities of type Message
+        Query messageQuery = new Query(MESSAGE_KIND).addSort(TIMESTAMP_PROPERTY, SortDirection.ASCENDING);
+        PreparedQuery preparedMessageQuery = datastore.prepare(messageQuery);
 
-    // Creates list of the messages text
-    ImmutableList<String> messagesList = Streams.stream(preparedMessageQuery.asIterable()).map(message -> 
+        ImmutableMap<String, ImmutableList<String>> data = getTemplateData(preparedMessageQuery);
+
+        final String out = getOutputString("chatPage", data);
+
+        response.getWriter().println(out);
+    }
+
+    /**
+     * Is called by a button on the chat form. doPost takes the message and its
+     * attributes (such as timestamp and user) and stores them into Datastore.
+     */
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        final String messageText = (String) request.getParameter(MESSAGE_TEXT_PROPERTY);
+        final long timestamp = System.currentTimeMillis();
+
+        // Reads API Key and Referer from file 
+        File apiKeyFile = new File(ROOT_FILE_PATH + "keys.txt");
+        Scanner scanner = new Scanner(apiKeyFile);
+        final String API_KEY = scanner.nextLine();
+        final String REFERER = scanner.nextLine();
+        scanner.close();
+
+        final String perspectiveURL = "https://commentanalyzer.googleapis.com/v1alpha1/comments:analyze?key=" + API_KEY;
+
+        final Double commentScore = getCommentScore(perspectiveURL, REFERER, messageText);
+
+        // If the comment is not toxic, then post it to Datastore
+        if (commentScore <= 0.85) {
+            DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+            Entity messageEntity = new Entity(MESSAGE_KIND);
+            messageEntity.setProperty(MESSAGE_TEXT_PROPERTY, messageText);
+            messageEntity.setProperty(TIMESTAMP_PROPERTY, timestamp);
+            datastore.put(messageEntity);
+
+            response.sendRedirect("/chat");
+        } else {
+            ImmutableMap<String, String> errorData = ImmutableMap.of("errorMessage", "Your message contains content that may " + 
+                "be deemed offensive by others. Please revise your message and try again.");
+
+            final String out = getOutputString("error", errorData);
+
+            response.getWriter().println(out);
+        }
+    }
+
+    /**
+     * Returns the output string for the response. In other words,
+     * it sets up the soy template with the passed in data.
+     */
+    private String getOutputString(String templateName, ImmutableMap data) {
```

Comment:
do you mean have the method take in three parameters `String templateName, String key, String value`? The only thing is that I call this in the doGet and in that case the value for the map is a list, so could I declare the value as type Object, so that different types could be passed in?

Reply:
ah I see that it's being passed in to the SoyTofu. I'll defer to Courtney since I'm not as familiar with this.

What I was trying to get at is that in all the use cases I see, the maps only have one type of key in it. Maps are useful for if you have multiple keys and need to be able to search to get the values for the different entries. If there's only ever one entry, the use of a map overly complex; there are simpler and clearer ways to pair the data together

Reply:
I think the maps are necessary here, since that's the data format that the renderer is expecting. 

However, since this code is a bit convoluted we should consider refactoring it so that it becomes more shareable and easier to understand. I think that we should create a new SoyRendererUtils class that includes all of this logic. Then each of the pages can pass in the template name and data so that all of the pages can share this, and this logic won't have to be repeated.